src/workerd/api/eventsource.h

```diff
@@ -0,0 +1,176 @@
+#pragma once
+#include <workerd/jsg/jsg.h>
+#include <workerd/jsg/url.h>
+#include "basics.h"
+
+namespace workerd::api {
+
+using kj::uint;
+class Fetcher;
+
+// Implements the web standard EventSource API.
+class EventSource: public EventTarget {
+public:
+  class ErrorEvent final: public Event {
+  public:
+    ErrorEvent(jsg::Lock& js, const jsg::JsValue& error)
+        : Event(kj::str("error")),
+          error(js, error) {}
+
+    static jsg::Ref<ErrorEvent> constructor() = delete;
+    JSG_RESOURCE_TYPE(ErrorEvent) {
+      JSG_INHERIT(Event);
+      JSG_LAZY_READONLY_INSTANCE_PROPERTY(error, getError);
+    }
+
+  private:
+    jsg::JsRef<jsg::JsValue> error;
+
+    jsg::JsValue getError(jsg::Lock& js) { return error.getHandle(js); }
+  };
+
+  class OpenEvent final: public Event {
+  public:
+    OpenEvent() : Event(kj::str("open")) {}
+    static jsg::Ref<ErrorEvent> constructor() = delete;
+    JSG_RESOURCE_TYPE(OpenEvent) {
+      JSG_INHERIT(Event);
+    }
+  };
+
+  class MessageEvent final: public Event {
+  public:
+    explicit MessageEvent(kj::Maybe<kj::String>& type,
+                 kj::String data,
+                 kj::Maybe<kj::String> lastEventId,
+                 jsg::Url& url)
+        : Event(kj::mv(type).orDefault([] { return kj::str("message");})),
+          data(kj::mv(data)),
+          lastEventId(kj::mv(lastEventId)),
+          origin(url.getOrigin()) {}
+
+    static jsg::Ref<ErrorEvent> constructor() = delete;
+    JSG_RESOURCE_TYPE(MessageEvent) {
+      JSG_INHERIT(Event);
+      JSG_LAZY_READONLY_INSTANCE_PROPERTY(data, getData);
+      JSG_LAZY_READONLY_INSTANCE_PROPERTY(origin, getOrigin);
+      JSG_LAZY_READONLY_INSTANCE_PROPERTY(lastEventId, getLastEventId);
+    }
+
+  private:
+    kj::String data;
+    kj::Maybe<kj::String> lastEventId;
+    kj::Array<const char> origin;
+
+    kj::StringPtr getData() { return data; }
+    kj::Maybe<kj::StringPtr> getLastEventId() {
+      return lastEventId.map([](kj::String& s) { return s.asPtr(); });
+    }
+    kj::ArrayPtr<const char> getOrigin() { return origin; }
+  };
+
+  struct EventSourceInit {
+    // We don't actually make use of the standard withCredentials option. If this is set to
+    // any truthy value, we'll throw.
+    jsg::Optional<bool> withCredentials;
```

Comment:
`RequestInitializerDict::credentials` is defined as `jsg::WontImplement`. I think the two should be consistent, but I don't have a strong opinion about which one should change.

Reply:
I think I prefer this approach ... that is, I think we should change the `jsg::WontImplement` in `RequestInitializerDict` and handle it equivalently as here. Don't have strong feelings on it tho.